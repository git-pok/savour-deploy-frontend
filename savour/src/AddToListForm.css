.AddToListForm-form {
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex-wrap: wrap;
  width: 70%;
  margin: 20px auto 40px auto;
  padding: 30px 0px;
  /* background-color: rgb(39, 55, 77); */
  background-color: rgb(78, 42, 42);
  border-radius: 10px;
}

.AddToListForm-form-field {
  display: block;
  width: 46%;
  margin: 10px 10px;
}

.AddToListForm-form-field label {
  display: block;
  width: 100%;
  height: 30px;
  margin: 0px auto;
  /* color: rgb(221, 230, 237); */
  color: rgb(255, 255, 255);
}

.AddToListForm-form-field input, .AddToListForm-form-field select {
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  padding: 0px 0px;
  margin: 0px auto 20px auto;
  border: 6px solid rgb(255, 255, 255);
  border-radius: 05px;
}

.AddToListForm-form-submit {
  display: block;
  width: 100%;
  margin-top: 10px;
}

.AddToListForm-form-submit button {
  display: block;
  width: 260px;
  height: 30px;
  margin: 0px auto;
  margin-top: 10px;
  border: 3px inset rgb(223, 216, 85);
  background-color: rgb(223, 216, 85);
  color: rgb(0, 0, 0);
  transition: .5s ease-in-out background-color;
}

.AddToListForm-form button:hover {
  border-color: rgb(255, 255, 255);
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
}

.AddToListForm-msg-div {
  display: block;
  margin: 0px auto;
}

.AddToListForm-msg-div div {
  margin: 0px auto;
}

.AddToListForm-spinner-icon {
  font-size: 10px;
  color: rgb(78, 42, 42);
  margin: 5px 5px;
}

@media only screen and (max-width: 1384px) and (min-width: 400px) {
  .AddToListForm-div {
    flex-direction: column;
    align-items: center;
    flex-wrap: nowrap;
  }

  .AddToListForm-form-field {
    width: 70%;
  }

  .AddToListForm-form-submit button {
    width: 50%;
    /* height: 100%; */
  }
}
