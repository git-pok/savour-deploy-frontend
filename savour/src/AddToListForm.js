import { useContext, useState, useEffect } from 'react';
import useAxios from './hooks/useAxios.js';
import UserContext from './context/UserContext.js';
import SavourApi from './models/SavourApi.js';
import useToggleState from './hooks/useToggleState.js';
import Message from './Message.js';
import './AddToListForm.css';
/**
 * AddToListForm
 * Props: recipelist, setState, setToggleState, recipelistSelectRecipe
 * Renders: renders form to add to recipe or shoppinglist.
*/
const AddToListForm = ({recipelist=false, setState=false, setToggleState=false, recipelistSelectRecipe=false}) => {
  // Destructure context.
  const { usrData, setUsrData } = useContext(UserContext);
  // Create req header for authen.
  const headers = { _token: `Bearer ${usrData.token}`};
  // Recipe request.
  const recipeOpts = { method: "get", url: `/recipes`, data: {}, params: {}, headers };
  const [ recipeData ] = useAxios(recipeOpts);
  // Occasion request.
  const OccOpts = { method: "get", url: `/occasions`, data: {}, params: {}, headers };
  const [ occasionData ] = useAxios(OccOpts);
  // Url for form request.
  const listReReqUrlEndPt = recipelist ? "recipelists" : "shoppinglists";
  const listReReqUrl = `/users/${usrData.userId}/${listReReqUrlEndPt}`;
  const recipelistUrl = `/users/${usrData.userId}/recipelists`;
  const recpLstOpts = {method: "get", url: recipelistUrl, data: {}, params: {}, headers};
  const [ recipelists, setRecipelists ] = useAxios(recpLstOpts);

  const [ listUrl, setListUrl ] = useState(null);
  const [ isRecipeList, setIsRecipeList ] = useToggleState(false);
  const [ isSubmitted, setIsSubmitted ] = useToggleState(false);

  const initialState = {
    shoplistName: "", recipelistName: "",
    occasionId: "", recipelistId: "",
    recipeId: recipelistSelectRecipe !== false ? recipelistSelectRecipe.id : ""
  };

  const [ formData, setFormData ] = useState(initialState);
  const [ formReqMade, setFormReqMade ] = useToggleState(false);
  const [ formErrMsg, setFormErrMsg ] = useState(null);
  const [ succMsg, setSuccMsg ] = useState(null);
  const [ isFormReqSucc, setIsFormReqSucc ] = useToggleState(false);

  /**
   * setReqUrl
   * args: shop
   * sets state with submitted form data,
   * which gets used in useEffect.
  */
  const setReqUrl = (shop=true) => {
    const userId = usrData.userId;
    // Check for occasion id from recipelist form data.
    const reclistId = formData.recipelistId;
    if (!shop && reclistId === "") {
      setFormErrMsg("Must select a list!");
      setFormReqMade();
      setTimeout(setFormReqMade, 3000);
      setTimeout(() => setFormErrMsg(null), 3000);
    } else {
      if (!shop) setListUrl(`/users/${userId}/recipelists/${reclistId}`);
      if (!shop) setIsRecipeList();
      else setListUrl(`/users/${userId}/shoppinglists`);
      setIsSubmitted();
    }
  }

  useEffect(() => {
    const createList = async () => {
      try {
        const data = isRecipeList ?
            { recipe_id: +formData.recipeId }
            : { recipe_id: +formData.recipeId, list_name: formData.shoplistName }

        const listReq = await SavourApi.request("post", listUrl, data, {}, headers);
        setSuccMsg(isRecipeList ? "Added item to list!" : "Created list for recipe!");
        setFormReqMade();
        setIsFormReqSucc();
        const reReq = await SavourApi.request("get", listReReqUrl, {}, {}, headers);

        if (setState) setTimeout(setState(() => [...reReq.data]), 3000);
        if (setToggleState) setTimeout(setToggleState, 3000);
        setTimeout(setIsSubmitted, 3000);
        if (isRecipeList) setTimeout(setIsRecipeList, 3000);
        setTimeout(setFormReqMade, 3000);
        setTimeout(setIsFormReqSucc, 3000);
        setTimeout(() => setSuccMsg(null), 3000);
        setTimeout(() => setListUrl(null), 3000);
      } catch(err) {
        console.log("ERROR!", err);
        const error = err.response.data.error.message;
        const isErrObj = typeof err.message === "object";
        const isErrorObj = typeof error === "object";
        const errMsg = isErrObj ? null : err.message;
        const errorMsg = isErrorObj ? null : error;
        setFormErrMsg(() => errorMsg || errMsg || "Error!");
        setFormReqMade();
        setTimeout(setIsSubmitted, 3000);
        if (isRecipeList) setTimeout(setIsRecipeList, 3000);
        setTimeout(setFormReqMade, 3000);
        setTimeout(() => setFormErrMsg(null), 3000);
        setTimeout(() => setListUrl(null), 3000);
      }
    }

    if (isSubmitted) createList();
  }, [isSubmitted])

  const handleChange = (evt) => {
    const { name, value } = evt.target;

    setFormData(data => ({
      ...data,
      [name]: value
    }))
  }

  const handleSubmit = async (evt) => {
    evt.preventDefault();
  }

  return (
    <>
    { !recipelist
      ?
        <form onSubmit={handleSubmit} className="AddToListForm-form">
        <div className="AddToListForm-form-field">
          <label htmlFor="shoplistName">List Name</label>
          <input
            id="shoplistName"
            type="text"
            name="shoplistName"
            placeholder="Type a list name"
            onChange={handleChange}
            value={formData.shoplistName} />
        </div>
        <div className="AddToListForm-form-field">
          <label htmlFor="recipeId">Recipes</label>
          <select
            id="recipeId"
            name="recipeId"
            onChange={handleChange}
            value={formData.recipeId}>
            <option key="select-a-recipe" value="">Select a Recipe</option>
            { recipeData &&
              recipeData.map(recipe => (
                <option key={`${recipe.id}`} value={`${recipe.id}`}>{recipe.name}</option>
              ))
            }
          </select>
        </div>
        <div className="AddToListForm-form-submit">
        <div className="AddToListForm-msg-div">
          { formReqMade &&
            <Message msgObj={
              {
                class: isFormReqSucc ? "success" : "fail",
                msg: isFormReqSucc ? succMsg : formErrMsg
              }
            } />
          }
        </div>
          <button onClick={setReqUrl}>CREATE LIST</button>
        </div>
        </form>
      :
        <form onSubmit={handleSubmit} className="AddToListForm-form">
          <div className="AddToListForm-form-field">
          <label htmlFor="reclistRecAdd">Recipelists</label>
          <select
            id="reclistRecAdd"
            name="recipelistId"
            onChange={handleChange}
            value={formData.recipelistId}>
            <option key="recipelistId" value="">Select a Recipelist</option>
              { recipelists &&
                recipelists.map(recipelist => (
                  <option key={`${recipelist.id}`} value={`${recipelist.id}`}>{recipelist.list_name}</option>
                ))
              }
          </select>
          </div>
          <div className="AddToListForm-form-field">
            <label htmlFor="reclistRecAdd">Recipes</label>
          <select
            id="recipeId"
            name="recipeId"
            onChange={handleChange}
            value={formData.recipeId}>
            { recipeData && recipelistSelectRecipe === false ?
              <option key="recipeId" value="">Select a Recipe</option>
            :
              <option key="recipeId" value={recipelistSelectRecipe.id}>{recipelistSelectRecipe.name}</option>
            }
            { recipeData && recipeData.length &&
              recipeData.map(recipe => (
                <option key={`${recipe.id}`} value={`${recipe.id}`}>{recipe.name}</option>
              ))
            }
          </select>
          </div>
          <div className="AddToListForm-form-submit">
          <div className="AddToListForm-msg-div">
            { formReqMade &&
              <Message msgObj={
                {
                  class: isFormReqSucc ? "success" : "fail",
                  msg: isFormReqSucc ? succMsg : formErrMsg
                }
              } />
            }
          </div>
          <button
            onClick={
              () => setReqUrl(false)
            }>ADD TO LIST</button>
          </div>
        </form>
    }
    </>
  );
}

export default AddToListForm;