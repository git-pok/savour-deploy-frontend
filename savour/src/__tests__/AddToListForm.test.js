import AddToListForm from '../AddToListForm.js';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../context/UserContext.js';

const usrData = {
  userUsername: "fvin",
  userId: 12
};

function renderAddToListForm () {
  return render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <AddToListForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
}

test("renders AddToListForm without crashing", () => {
  renderAddToListForm();
});

test("renders snapshot without crashing", () => {
  const { asFragment } = renderAddToListForm();
  expect(asFragment()).toMatchSnapshot();
});