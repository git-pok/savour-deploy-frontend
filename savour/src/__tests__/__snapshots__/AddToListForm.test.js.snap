// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <form
    class="AddToListForm-form"
  >
    <div
      class="AddToListForm-form-field"
    >
      <label
        for="shoplistName"
      >
        List Name
      </label>
      <input
        id="shoplistName"
        name="shoplistName"
        placeholder="Type a list name"
        type="text"
        value=""
      />
    </div>
    <div
      class="AddToListForm-form-field"
    >
      <label
        for="recipeId"
      >
        Recipes
      </label>
      <select
        id="recipeId"
        name="recipeId"
      >
        <option
          value=""
        >
          Select a Recipe
        </option>
      </select>
    </div>
    <div
      class="AddToListForm-form-submit"
    >
      <div
        class="AddToListForm-msg-div"
      />
      <button>
        CREATE LIST
      </button>
    </div>
  </form>
</DocumentFragment>
`;
